tests/test_build.py

```diff
@@ -254,3 +255,96 @@ def test_included_toc(sphinx_build_factory):
     sphinx_build = sphinx_build_factory("test_included_toc").build()
     included_page_html = sphinx_build.html_tree("included-page.html")
     assert included_page_html is not None
+
+
+good_edits = [
+    [
+        {
+            "github_user": "foo",
+            "github_repo": "bar",
+            "github_version": "HEAD",
+            "doc_path": "docs",
+        },
+        "https://github.com/foo/bar/edit/HEAD/docs/index.rst",
+    ],
+    [
+        {
+            "gitlab_user": "foo",
+            "gitlab_repo": "bar",
+            "gitlab_version": "HEAD",
+            "doc_path": "docs",
+        },
+        "https://gitlab.com/foo/bar/-/edit/HEAD/docs/index.rst",
+    ],
+    [
+        {
+            "bitbucket_user": "foo",
+            "bitbucket_repo": "bar",
+            "bitbucket_version": "HEAD",
+            "doc_path": "docs",
+        },
+        "https://bitbucket.org/foo/bar/src/HEAD/docs/index.rst?mode=edit",
+    ],
+]
+
+slash_edits = [
+    [{k: f"{v}/" if k == "doc_path" else v for k, v in ctx.items()}, url]
+    for ctx, url in good_edits
+]
+
+providers = [
+    [
+        dict(**ctx, **{f"{provider}_url": f"https://{provider}.example.com"}),
+        f"""https://{provider}.example.com/foo/{url.split("/foo/")[1]}""",
+    ]
+    for ctx, url in good_edits
+    for provider in ["gitlab", "bitbucket", "github"]
+    if provider in url
+]
+
+bad_edits = [
```

Comment:
Can you add some brief comment to each of those lists that explains the content? (the dict comprehension is a bit hard to understand from a glance, you already need to know what's inside the dicts, and that there is a key with "version" in it that gets skipped in this one)